FirebaseMLModelDownloader/Sources/ModelInfoRetriever.swift

```diff
@@ -281,59 +275,122 @@ class ModelInfoRetriever {
+                  .invalidArgument(self.modelName)
+                DeviceLogger.logEvent(level: .debug,
+                                      message: description,
+                                      messageCode: .invalidArgument)
+              }
               self.telemetryLogger?.logModelInfoRetrievalEvent(eventName: .modelDownload,
-                                                               status: .failed,
+                                                               status: .modelInfoRetrievalFailed,
                                                                errorCode: .httpError(code: httpResponse
                                                                  .statusCode))
               completion(.failure(.invalidArgument))
             case 401, 403:
-              DeviceLogger.logEvent(level: .debug,
-                                    message: ModelInfoRetriever.ErrorDescription.permissionDenied,
-                                    messageCode: .permissionDenied)
+              // Error could be due to FirebaseML API not enabled for project, or invalid permissions.
+              if let data = data,
+                let responseJSON = try? JSONSerialization
+                .jsonObject(with: data, options: []) as? [String: Any],
+                let error = responseJSON["error"] as? [String: Any],
+                let errorMessage = error["message"] as? String {
+                DeviceLogger.logEvent(level: .debug,
+                                      message: errorMessage,
+                                      messageCode: .permissionDenied)
+              } else {
+                DeviceLogger.logEvent(level: .debug,
+                                      message: ModelInfoRetriever.ErrorDescription.permissionDenied,
+                                      messageCode: .permissionDenied)
+              }
               self.telemetryLogger?.logModelInfoRetrievalEvent(eventName: .modelDownload,
-                                                               status: .failed,
+                                                               status: .modelInfoRetrievalFailed,
                                                                errorCode: .httpError(code: httpResponse
                                                                  .statusCode))
               completion(.failure(.permissionDenied))
             case 404:
-              let description = ModelInfoRetriever.ErrorDescription.modelNotFound(self.modelName)
-              DeviceLogger.logEvent(level: .debug,
-                                    message: description,
-                                    messageCode: .modelNotFound)
+              if let data = data,
+                let responseJSON = try? JSONSerialization
+                .jsonObject(with: data, options: []) as? [String: Any],
+                let error = responseJSON["error"] as? [String: Any],
+                let errorMessage = error["message"] as? String {
```

Comment:
We still have a room to reduce code duplication and make it more tolerant to potential response format changes, but looks fine for now.